Fix #6866: Remove webpack compilation from backend tests

ankita240796 opened the pull request:
## Explanation
Fixes #6866: Adds a new path for frontend templates while running backend tests to avoid webpack compilation before running the tests.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

ankita240796 committed:
Remove compilation from backend tests

vojtechjelinek commented on assets/constants.ts:
```diff
@@ -612,5 +612,8 @@ export = {
+  "DEV_MODE": true,
+
+  // Whether backend tests are running.
+  "RUN_BACKEND_TESTS": false
```
I'm not sure if it is good practice to use this file for saving temporary variables, also the variable is only used in backend so it should be probably moved to `feconf.py`.

ankita240796 replied:
Updated.

ankita240796 committed:
Fix failing test

vojtechjelinek reviewed (changes requested):
It seems to me that most of these test should actually be e2e tests because they test both backend and frontend. (or right now they only test that we serve a static page with some specified content).

ankita240796 committed:
Move var to feconf, make coverage 100%

ankita240796 commented:
Hi @vojtechjelinek, I have moved the temporary var to feconf file. Also regarding your suggestion here:

> It seems to me that most of these test should actually be e2e tests because they test both backend and frontend. (or right now they only test that we serve a static page with some specified content).

I have already updated the tests which were failing to check for some specific content in the static page before being served. 

PTAL, Thanks!

brianrodri commented on core/controllers/base_test.py:
```diff
@@ -104,6 +105,21 @@ def setUp(self):
+            pass
+        init_swap = self.swap(base.BaseHandler, '__init__', mock_init)
+        with init_swap, self.swap(feconf, 'RUN_BACKEND_TESTS', False):
+            env = base.BaseHandler('req', 'res').jinja2_env
```
Please add a comment explaining that `jinja2_env` is a `@webapp2.cached_property`, I was surprised this test worked and had to look it up myself.

ankita240796 replied:
Done.

brianrodri commented on core/controllers/email_dashboard_test.py:
```diff
@@ -156,7 +156,8 @@ def test_email_dashboard_page(self):
         response = self.get_html_response('/emaildashboard')
         self.assertIn(
-            '<title itemprop="name">Email Dashboard - Oppia</title>',
+            '@require(\'../../base-components/header.template.html\', '
```
Why add checks for `@require` in some tests and not others?

ankita240796 replied:
These tests are present from before and they check that the html response obtained contains a line specific to the page. Since we are not compiling webpack, we don't have the pages in the serving state, so I am just comparing the require line which after compilation should change into the line which was being tested earlier.

brianrodri replied:
OK, fair enough!

brianrodri commented on jinja_utils.py:
```diff
@@ -86,6 +86,24 @@ def get_jinja_env(dir_path):
 
 
+def get_jinja_env_for_backend_tests(template_root_dir):
+    subdirs = []
```
Maybe do a `self.swap` here as well just in case someone manually sets the feconf value to false?

ankita240796 replied:
I am sorry, I did not understand this query. We are changing the value `RUN_BACKEND_TEST` to True before starting the backend tests and this function will be called only if the value is True, so even if we swap the value inside the function, it will be of no use because it won't be called unless the value is True.

brianrodri replied:
Ah I see, I missed that. Nevermind!

brianrodri commented on scripts/run_backend_tests.py:
```diff
@@ -348,148 +358,154 @@ def main(args=None):
-        tasks.append(task)
-
-    task_execution_failed = False
+    update_feconf('RUN_BACKEND_TESTS = True')
```
Hmm, why do we have to change the file directly? Can't we just "permanently" swap the value to True using pure Python?

ankita240796 replied:
We will need to do a swap in multiple test files which have tests which require fetching a template, so it is better to change it before starting all the tests.

brianrodri replied:
Right, I'm saying to "swap", but never "swap" back. Basically, look into the "swap" implementation and do the first step, but don't let it return to its original value.

ankita240796 replied:
Can you please guide me on how can I do this?

Should I create a function as follows:

```
def update_feconf(value):
    """Updates feconf file to set the value RUN_BACKEND_TESTS.

    Args:
        value: bool. True/False value for RUN_BACKEND_TESTS.
    """
    setattr(feconf, 'RUN_BACKEND_TESTS', value)
```

brianrodri replied:
Yup, that looks right to me. This will allow us to avoid making file-changes which should make things much easier on developers.

ankita240796 replied:
I tried adding a swap function as follows:

```
@contextlib.contextmanager
def swap_feconf_flag(newvalue):
    """Updates feconf file to set the value RUN_BACKEND_TESTS.

    Args:
        newvalue: bool. True/False value for RUN_BACKEND_TESTS.
    """
    original = getattr(feconf, 'RUN_BACKEND_TESTS')
    setattr(feconf, 'RUN_BACKEND_TESTS', newvalue)
    try:
        yield
    finally:
        setattr(feconf, 'RUN_BACKEND_TESTS', original)
```

I moved the complete code inside try block to with block:

```
with swap_feconf_flag(True):
```

If I print the value of flag inside with block, it is True but inside tests it shows up as False and the test fails. Am I doing any part wrong here?

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

ankita240796 committed:
Add comment

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

seanlip commented on core/controllers/base.py:
```diff
@@ -129,6 +129,9 @@ def jinja2_env(self):
         Returns:
             Environment. A Jinja2 environment object used to load templates.
         """
+        if feconf.RUN_BACKEND_TESTS:
```
I'm not sure about this -- I believe Jinja is on the verge of being removed entirely, so it may not make sense to introduce a flag like this in the main codebase. @vojtechjelinek what do you think?

vojtechjelinek replied:
I mean the issue is that we should either replace the tests that check for some HTML content with e2e tests OR test on the files that are built by webpack so that we are as close to the real production situation as possible (an don't introduce a skew).

Because some of the backend tests are more like meta tests that check if some data are consistent. For example [topics_and_skills_dashboard_test.py#L142](https://github.com/oppia/oppia/blob/b1b102cce8c9d1eda1a1872c67b97748a84ff8f8/core/controllers/topics_and_skills_dashboard_test.py#L142) checks that some handlers serve at least something and that it cointains `require(\'../../base-components/header.template.html\', {"title": "Topics and Skills Dashboard - Oppia"})` which doesn't really check anything important (that can't be checked with e2e tests).

vojtechjelinek replied:
I think that for handlers that are static we should trust e2e tests and for handlers that have some kind of permissions we shouldn't check the actual content returned but only if some permission error was fired or not.

ankita240796 replied:
We can remove the checking of `require` part from the tests but we will still need the flag since when `get_template` function is called for obtaining signup page or any other page, it looks inside webpack_bundles (that is used as file loader by jinja_env). If we want to skip webpack compilation, we need to have separate file loader for test to serve these pages. So, we need a flag to choose which loader we want to use.

brianrodri commented on scripts/run_backend_tests.py:
```diff
@@ -348,148 +358,154 @@ def main(args=None):
+    update_feconf('RUN_BACKEND_TESTS = True')
+    # This complete block is put in try...finally to enusre that update made
+    # to feconf file is reverted even in case of an exception.
     try:
```
You could also replace this `try/except` with `with swap(feconf, 'RUN_BACKEND_TESTS', True):`

ankita240796 replied:
I updated this but the it does not work correctly for test. I have described the issue in this [comment](https://github.com/oppia/oppia/pull/7604#discussion_r324033168).

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

ankita240796 committed:
Fix lint issues

ankita240796 committed:
Revert changes

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

ankita240796 committed:
Revert changes

ankita240796 committed:
Revert changes

ankita240796 committed:
Add swap to test utils

ankita240796 committed:
Fix test

ankita240796 commented:
@seanlip suggested to have a mock defined in `test_utils.py` for obtaining a template instead of adding a new mode for backend tests to ensure that the tests are working in the same environment as the codebase. I have updated the PR according to this suggestion, PTAL. Thanks!

seanlip commented on core/domain/interaction_registry_test.py:
```diff
@@ -55,10 +55,11 @@ def test_dependency_loads_in_exploration_player_page(self):
-        response.mustcontain('skulpt')
+        response.mustcontain(
+            '@require(\'../../../../../../extensions/interactions/'
+            'dependency_html.html\')')
```
You should update line 66 similarly.

ankita240796 replied:
Done.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -80,6 +81,51 @@ def empty_environ():
 
 
+def get_filepath_from_filename(filename, rootdir):
+    """Returns filepath using the filename.
```
This needs way more explanation. What's the context? Why not just do os.path.join(rootdir, filename)? Please explain what is going on more, so that developers who have no idea what this function does can understand it based on the docstring.

ankita240796 replied:
Updated.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -80,6 +81,51 @@ def empty_environ():
+        for name in files:
+            if name == filename:
+                return os.path.join(root, filename)
+    return ''
```
None is more semantically correct. (Update docstring typeinfo accordingly.)

ankita240796 replied:
Updated.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -80,6 +81,51 @@ def empty_environ():
+
+
+def mock_get_template(unused_self, filename):
+    """Mock for get_template function of jinja2 Environment.
```
Explain reasoning behind having this function, etc.

ankita240796 replied:
Done.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -546,9 +592,11 @@ def _get_response(
-        response = self.testapp.get(
-            url, params, expect_errors=expect_errors,
-            status=expected_status_int)
+        with self.swap(
```
Explain why you are doing this, here and below.

ankita240796 replied:
Done.

seanlip reviewed (changes requested):
Thanks @ankita240796 -- the overall approach seems reasonable. Left a few inline comments, PTAL!

ankita240796 committed:
Add comments

ankita240796 commented:
Hi @seanlip, I have addressed the review comments, PTAL. Thanks!

jameesjohn commented on core/controllers/email_dashboard_test.py:
```diff
@@ -156,7 +156,8 @@ def test_email_dashboard_page(self):
         self.assertIn(
-            '<title itemprop="name">Email Dashboard - Oppia</title>',
+            '@require(\'../../base-components/header.template.html\', '
+            '{"title": "Email Dashboard - Oppia"})',
```
I think it will be best to just check for `{"title": "Email Dashboard - Oppia"}` instead of the whole require statement

ankita240796 replied:
Updated.

jameesjohn commented on core/controllers/email_dashboard_test.py:
```diff
@@ -216,7 +217,8 @@ def test_email_dashboard_result_page(self):
         self.assertIn(
-            '<title itemprop="name">Email Dashboard Result - Oppia</title>',
+            '@require(\'../../base-components/header.template.html\', '
+            '{"title": "Email Dashboard Result - Oppia"})',
```
Ditto here

ankita240796 replied:
Updated.

jameesjohn commented on core/controllers/learner_dashboard_test.py:
```diff
@@ -284,7 +284,8 @@ def test_learner_dashboard_page(self):
         self.assertIn(
-            '<title itemprop="name">Learner Dashboard - Oppia</title>',
+            '@require(\'../../base-components/header.template.html\', '
+            '{"title": "Learner Dashboard - Oppia"})',
```
Here as well as every other place that it is required

ankita240796 replied:
Updated.

jameesjohn reviewed (changes requested):
Thanks @ankita240796.
I left a few comments PTAL

ankita240796 committed:
Remove extra require from tests

ankita240796 commented:
Hi @Jamesjay4199, I have addressed the review comments, PTAL. Thanks!

vojtechjelinek commented on core/tests/test_utils.py:
```diff
@@ -80,6 +81,68 @@ def empty_environ():
+    filename by using os.path.join(root, filename).
+
+    For example signup-page.mainpage.html is present in
+    core/templates/dev/head/pages/signup-page and error-pgae.mainpage.html is
```
`error-pgae.mainpage.html` -> `error-page.mainpage.html`

ankita240796 replied:
Done.

vojtechjelinek commented on core/tests/test_utils.py:
```diff
@@ -658,9 +733,16 @@ def get_json(self, url, params=None, expected_status_int=200):
+            jinja2.environment.Environment, 'get_template', mock_get_template):
+            json_response = self.testapp.get(
+                url, params, expect_errors=expect_errors,
+                status=expected_status_int)
```
Do we really need to swap for JSON response?

ankita240796 replied:
Yes, my bad. It is not required here. Removed.

vojtechjelinek reviewed (approved):
I really like the approach with mocking the method, just a fe questions.

I still think that we should get rid of the tests that check for some content inside the HTML (like `<exploration-editor-page></exploration-editor-page>`) but I think that we should dicuss this in separate issue.

jameesjohn reviewed (approved):
LGTM!

ankita240796 committed:
Fix spelling

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

ankita240796 committed:
Remove extra swap

ankita240796 committed:
Remove extra space

nithusha21 reviewed (approved):
I think this LGTM. No comments from my side.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -80,6 +81,68 @@ def empty_environ():
+    filepath = get_filepath_from_filename(
+        filename, os.path.join('core', 'templates', 'dev', 'head', 'pages'))
+    if not filepath:
+        filepath = get_filepath_from_filename(
```
I understand this mock retrieving from /pages. But why is it retrieving from /tests as well? That part seems weird.

ankita240796 replied:
This is required since we have the `console_errors.html` template in `core/templates/dev/head/tests` folder. I have added a comment as well.

seanlip replied:
Ah OK. Don't do it this way. /console_errors path is only needed for a protractor test, and the backend tests don't need this, right? There should be just one canonical directory for retrieving templates, so ideally get rid of these three lines of code altogether.

ankita240796 replied:
The test [here](https://github.com/oppia/oppia/blob/develop/core/controllers/base_test.py#L107) checks this url.

seanlip replied:
Exclude this URL explicitly from the test you linked (and explain why).

ankita240796 replied:
Done.

seanlip commented on core/tests/test_utils.py:
```diff
@@ -80,6 +81,68 @@ def empty_environ():
         os.environ['HTTP_HOST'], os.environ['SERVER_PORT'])
 
 
+def get_filepath_from_filename(filename, rootdir):
```
What if (possibly in the future) two files have the same filename? Maybe detect that case and throw exception? Or have separate backend test that verifies that all files have different filenames (probably better since that means you can reference that test from here, and break early in the for loop here)

ankita240796 replied:
I can add a backend test but I have two doubts:

- Is it necessary for files to have distinct names always? The test may fail if we later have any case which requires two same filenames. (Though this is rare and we can ignore this)
- The test will import the `test_utils` file so I cannot it call it back in `test_utils` since that will create a circular dependency.

seanlip replied:
Probably a separate backend test would be better then, due to the circular dependency. And I think you can assume different filenames for now, and if they need to be the same, we can handle that later. (Better to add a check now than to ignore that case and have silent failures or unexpected behaviour later.)

ankita240796 replied:
I am sorry I don't clearly understand what you mean by separate backend test. Do we need to add a test which does not have class inheritance from `test_utils`?

seanlip replied:
It should be a regular backend test, but I don't see why you'd need to call it from test_utils.

ankita240796 replied:
Yes, it is fine if I do not call it but you suggested earlier:

```
probably better since that means you can reference that test from here, and break early in the for loop here
```

So, should I write part to check for same names in some common utility file?

seanlip replied:
Oh -- "reference" in the original note just means "leave a comment".

You can write this in base_test.py (but add clear documentation to explain the context).

ankita240796 replied:
Done.

seanlip reviewed (changes requested):
Just a couple more notes, thanks!

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

ankita240796 committed:
Explain tests directory usage

ankita240796 commented:
Hi @seanlip, PTAL. Thanks!

DubeySandeep reviewed (approved):
LGTM for the code owner file:
`core/controllers/editor_test.py`

kevinlee12 reviewed (approved):
lgtm!

ankita240796 committed:
Remove console_errors template

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

ankita240796 committed:
Add test for unique filenames

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

ankita240796 committed:
Fix lint

ankita240796 commented:
Hi @seanlip, PTAL. Thanks!

seanlip reviewed (approved):
Looks good. Thanks!

seanlip commented:
Hi @aks681 -- could you PTAL as codeowner so that this can be merged? Thanks!

aks681 reviewed (approved):
lgtm as codeowner
